_posts/2020-06-18-Metal3-dev-env-BareMetal-Cluster-Deployment.md

```diff
@@ -0,0 +1,392 @@
+  template:
+    metadata:
+      labels:
+        cluster.x-k8s.io/cluster-name: bmetalcluster
+        nodepool: nodepool-0
+    spec:
+      bootstrap:
+        configRef:
+          apiVersion: bootstrap.cluster.x-k8s.io/v1alpha3
+          kind: KubeadmConfigTemplate
+          name: bmetalcluster-workers
+      clusterName: bmetalcluster
+      infrastructureRef:
+        apiVersion: infrastructure.cluster.x-k8s.io/v1alpha3
+        kind: Metal3MachineTemplate
+        name: bmetalcluster-workers
+      version: v1.18.0
+status:
+  observedGeneration: 1
+  phase: ScalingUp
+  replicas: 1
+  selector: cluster.x-k8s.io/cluster-name=bmetalcluster,nodepool=nodepool-0
+  unavailableReplicas: 1
+  updatedReplicas: 1
+```
+
+<br/>
+To check the status we can follow steps similar to Controlplane case :
+
+```console
+kubectl get bmh -n metal3 -w
+```
+> We can see the live status of the node being provisioned. As mentioned before `bmh` is short representation for `BareMetalHosts`.
+
+```console
+kubectl get Machine -n metal3 -w
+```
+> This shows the status of Machines associated with workers, apart from the one for Controlplane, and we can watch the status of provisioning under PHASE 
+
+```console
+sudo virsh net-dhcp-leases baremetal
+```
+> To get the nodes IP
+
+```console
+ssh metal3@{control-plane-node-ip}
+kubectl get nodes
+```
+> To check if its added to the cluster
+
+```console
+ssh metal3@{node-ip}
+```   
+> If you want to login to the node
+
+```console
+kubectl scale --replicas=3 MachineDeployment ${CLUSTER_NAME} -n metal3
+```
+> We can add or remove workers to the cluster, we can scale up the MachineDeployment up or down, in this example we are adding 2 more worker nodes, making the total nodes = 3
+
+<br/>
+<br/>
+
+### Deprovisioning
+
+All of the previous components have corresponding deprovisioning scripts which use config files, in the previously mentioned manifest directory, and use them to clean up worker, control plane and cluster.
+
+This step will use the already generated cluster/control plane/worker definition file, and supply it to **Kubernetes** ansible module to remove/deprovision the resource. You can find it, under the `Manifest` directory, in the Snapshot shared at the beginning of this blogpost where we show the file structure.
+
+For example if you wish to deprovision the cluster, you would do :
+
+```console
+sh ${metal3-dev-env-path}/scripts/v1alphaX/deprovision_worker.sh
+sh ${metal3-dev-env-path}/scripts/v1alphaX/deprovision_controlplane.sh
+sh ${metal3-dev-env-path}/scripts/v1alphaX/deprovision_cluster.sh
+```
+> Note :
+> The reason for running the `deprovision_worker.sh` and `deprovision_controlplane.sh` scripts is that not all objects are cleared when we just run the `deprovision_cluster.sh` script. Following this, if you want to deprovision control plane it is recommended to run the scripts for both control plane and worker. For worker deprovisioning, we only need to run the worker script.
```

Comment:
I am not sure it makes much sense to deprovision the control plane without deprovisioning the cluster, as you won't be able to provision a new control plane with the same cluster, if the initial control plane is gone. so I would say that the workers can be deprovisioned on their own, but if you want to deprovision the control plane you should also do the cluster itself.

Reply:
Makes sense.